### PR TITLE
added CurveSwapper

### PR DESCRIPTION
**Pull request by vpomo**
added CurveSwapper

### CODE REVIEW

File: contracts/swappers/CurveSwapper.sol [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+    ICurveMinter curveMinter = ICurveMinter(minter);
+
+    uint amountIn = IERC20(tokenIn).balanceOf(address(this));
+
+    (uint256 tokenInIndex, uint256 tokenOutIndex) = getTokensIndex(minter, tokenIn, tokenOut);
+
+    // scope for checking price impact
+    uint amountOutMax;
+    {
+      uint minimalAmount = amountIn / 1000;
```
Comment (belbix) [inline contracts/swappers/CurveSwapper.sol:161]: is it necessary for curve?

File: contracts/swappers/CurveSwapper.sol [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+      amountOutMax = price * amountIn / minimalAmount;
+    }
+
+    require(IERC20(tokenIn).balanceOf(address(this)) >= amountIn,
+      "Wrong amountIn"
+    );
+    IERC20(tokenIn).approve(minter, amountIn);
+
+    uint amountOut = curveMinter.exchange(tokenInIndex, tokenOutIndex, amountIn, _LIMIT);
+    require(IERC20(tokenOut).balanceOf(address(this)) >= amountOut,
```
Comment (belbix) [inline contracts/swappers/CurveSwapper.sol:173]: inline

File: contracts/swappers/CurveSwapper.sol [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+      "Wrong amountIn"
+    );
+
+    require(amountOutMax < amountOut ||
+      (amountOutMax - amountOut) * PRICE_IMPACT_DENOMINATOR / amountOutMax <= priceImpactTolerance,
+      "!PRICE"
+    );
+
+    IERC20(tokenOut).transfer(recipient, amountOut);
```
Comment (belbix) [inline contracts/swappers/CurveSwapper.sol:182]: safe transfer

File: contracts/swappers/CurveSwapper.sol [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+        tokenOutIndex = i;
+        break;
+      }
+    }
+
+    require(tokenInIndex < len, 'Wrong tokenIn');
+    require(tokenOutIndex < len, 'Wrong tokenOut');
+  }
+
+  function getTokensFromMinter(address minter) private view returns(address[] memory) {
```
[the remark this answers is not in the document]
Comment (vpomo) [reply contracts/swappers/CurveSwapper.sol:107]: added

File: contracts/swappers/CurveSwapper.sol [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+      uint minimalAmount = amountIn / 1000;
+      require(minimalAmount != 0, "Too low amountIn");
+      uint price = getPrice(minter, tokenIn, tokenOut, minimalAmount);
+      amountOutMax = price * amountIn / minimalAmount;
+    }
+
+    require(IERC20(tokenIn).balanceOf(address(this)) >= amountIn,
+      "Wrong amountIn"
+    );
+    IERC20(tokenIn).approve(minter, amountIn);
```
[the remark this answers is not in the document]
Comment (vpomo) [reply contracts/swappers/CurveSwapper.sol:170]: transferFrom(msg.sender, .... is used there

File: contracts/swappers/CurveSwapper.sol (hunk 2 above)
Comment (vpomo) [reply contracts/swappers/CurveSwapper.sol:173]: removed

File: contracts/swappers/CurveSwapper.sol [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+    // scope for checking price impact
+    uint amountOutMax;
+    {
+      uint minimalAmount = amountIn / 1000;
+      require(minimalAmount != 0, "Too low amountIn");
+      uint price = getPrice(minter, tokenIn, tokenOut, minimalAmount);
+      amountOutMax = price * amountIn / minimalAmount;
+    }
+
+    require(IERC20(tokenIn).balanceOf(address(this)) >= amountIn,
```
[the remark this answers is not in the document]
Comment (vpomo) [reply contracts/swappers/CurveSwapper.sol:167]: removed

File: contracts/swappers/CurveSwapper.sol [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+
+    (uint256 tokenInIndex, uint256 tokenOutIndex) = getTokensIndex(minter, tokenIn, tokenOut);
+
+    // scope for checking price impact
+    uint amountOutMax;
+    {
+      uint minimalAmount = amountIn / 1000;
+      require(minimalAmount != 0, "Too low amountIn");
+      uint price = getPrice(minter, tokenIn, tokenOut, minimalAmount);
+      amountOutMax = price * amountIn / minimalAmount;
```
[the remark this answers is not in the document]
Comment (vpomo) [reply contracts/swappers/CurveSwapper.sol:164]: changed the logic

File: contracts/swappers/CurveSwapper.sol (hunk 1 above)
Comment (vpomo) [reply contracts/swappers/CurveSwapper.sol:161]: changed the logic

File: contracts/swappers/CurveSwapper.sol [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+
+    if (uintSuccess) {
+      amountOut = abi.decode(uintReturnData,(uint256));
+    } else {
+      ( , bytes memory intReturnData) = minter.call(
+        abi.encodeWithSignature("exchange(int128,int128,uint256,uint256)",
+        _convertToInt(tokenInIndex), _convertToInt(tokenOutIndex), dx, minDy
+        )
+      );
+      amountOut = abi.decode(intReturnData,(uint256));
```
Comment (belbix) [inline contracts/swappers/CurveSwapper.sol:254]: i'm not sure but looks like you convert int value to uint without cheking

File: contracts/swappers/CurveSwapper.sol [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+      );
+      dy = abi.decode(intReturnData,(uint256));
+    }
+  }
+
+  function _callExchange(
+    address minter, uint256 tokenInIndex, uint256 tokenOutIndex, uint256 dx, uint256 minDy
+  ) private returns (uint256 amountOut) {
+
+    (bool uintSuccess, bytes memory uintReturnData) = minter.call(
```
Comment (belbix) [inline contracts/swappers/CurveSwapper.sol:240]: is it necessary have two calls?

File: contracts/swappers/CurveSwapper.sol [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+    // scope for checking price impact
+    uint256 priceBefore = getPrice(pool, tokenIn, tokenOut, amountIn);
+
+    require(IERC20(tokenIn).balanceOf(address(this)) >= amountIn, "Wrong amountIn");
+    IERC20(tokenIn).approve(curveMinter, amountIn);
+
+    uint256 amountOut = _callExchange(curveMinter, tokenInIndex, tokenOutIndex, amountIn, _LIMIT);
+    require(IERC20(tokenOut).balanceOf(address(this)) >= amountOut, "Wrong amountOut");
+
+    uint256 priceAfter = getPrice(pool, tokenIn, tokenOut, amountIn);
```
Comment (belbix) [inline contracts/swappers/CurveSwapper.sol:109]: 1) getPrice pretty expensive, no need to call it twice
2) you need to check what you exactly received, calculate the price from received amounts

File: contracts/swappers/CurveSwapper.sol [hunk 11]
```diff
@@ -0,0 +1,257 @@
...
+    (uint256 tokenInIndex, uint256 tokenOutIndex) = getTokensIndex(curveMinter, tokenIn, tokenOut);
+
+    // scope for checking price impact
+    uint256 priceBefore = getPrice(pool, tokenIn, tokenOut, amountIn);
+
+    require(IERC20(tokenIn).balanceOf(address(this)) >= amountIn, "Wrong amountIn");
+    IERC20(tokenIn).approve(curveMinter, amountIn);
+
+    uint256 amountOut = _callExchange(curveMinter, tokenInIndex, tokenOutIndex, amountIn, _LIMIT);
+    require(IERC20(tokenOut).balanceOf(address(this)) >= amountOut, "Wrong amountOut");
```
Comment (belbix) [inline contracts/swappers/CurveSwapper.sol:107]: you check balance here without strict == and then transfer `amountOut`

**Review by belbix — CHANGES_REQUESTED** [review]

File: contracts/swappers/CurveSwapper.sol (hunk 8 above)
Comment (vpomo) [reply contracts/swappers/CurveSwapper.sol:254]: i am checking in function _convertToInt

File: contracts/swappers/CurveSwapper.sol (hunk 9 above)
Comment (vpomo) [reply contracts/swappers/CurveSwapper.sol:240]: yes
we have two types of exchange function:
1. "exchange(uint256,uint256,uint256,uint256)"
2. "exchange(int128,int128,uint256,uint256)"

File: contracts/swappers/CurveSwapper.sol [hunk 12]
```diff
@@ -0,0 +1,257 @@
...
+      dy = abi.decode(intReturnData,(uint256));
+    }
+  }
+
+  function _callExchange(
+    address minter, uint256 tokenInIndex, uint256 tokenOutIndex, uint256 dx, uint256 minDy
+  ) private returns (uint256 amountOut) {
+
+    (bool uintSuccess, bytes memory uintReturnData) = minter.call(
+      abi.encodeWithSignature("exchange(uint256,uint256,uint256,uint256)",
```
[the remark this answers is not in the document]
Comment (vpomo) [reply contracts/swappers/CurveSwapper.sol:241]: I would like to make a universal solution for different types of pools
we will have to store pool addresses to use interfaces

File: contracts/swappers/CurveSwapper.sol [hunk 13]
```diff
@@ -0,0 +1,257 @@
...
+    address curveMinter = _getMinter(pool);
+    uint amountIn = IERC20(tokenIn).balanceOf(address(this));
+
+    (uint256 tokenInIndex, uint256 tokenOutIndex) = getTokensIndex(curveMinter, tokenIn, tokenOut);
+
+    // scope for checking price impact
+    uint256 priceBefore = getPrice(pool, tokenIn, tokenOut, amountIn);
+
+    require(IERC20(tokenIn).balanceOf(address(this)) >= amountIn, "Wrong amountIn");
+    IERC20(tokenIn).approve(curveMinter, amountIn);
```
[the remark this answers is not in the document]
Comment (vpomo) [reply contracts/swappers/CurveSwapper.sol]: used

File: contracts/swappers/CurveSwapper.sol (hunk 10 above)
Comment (vpomo) [reply contracts/swappers/CurveSwapper.sol]: fixed

File: contracts/swappers/CurveSwapper.sol [hunk 14]
```diff
@@ -0,0 +1,270 @@
...
+  /// @param priceImpactTolerance Price impact tolerance. Must include fees at least.
+  function swap(
+    address pool,
+    address tokenIn,
+    address tokenOut,
+    address recipient,
+    uint priceImpactTolerance
+  ) external override {
+    address curveMinter = _getMinter(pool);
+    uint amountIn = IERC20(tokenIn).balanceOf(address(this));
```
Comment (belbix) [inline contracts/swappers/CurveSwapper.sol:96]: add require !=0

Comment (vpomo) [reply contracts/swappers/CurveSwapper.sol]: added

**Pull request merged**